src/MathOptInterface.jl

```diff
@@ -5,6 +5,29 @@ module MathOptInterface
 
 Abstract supertype for objects that implement the "Model" interface for defining
 an optimization problem.
+
+## Interface
+
+The API of MathOptInterface is large. At a minimum, objects subtyping
+`ModelLike` must implement [`get`](@ref) for the following attributes:
```

Comment:
Again, only if you want to be readable

Reply:
No. I think it's _all_. If they aren't readable, they won't implement `get` for the specific attributes. I think we can ask all models "how many constraints do you have" regardless of whether we can access them.

Reply:
I'm not going to implement `NumberOfConstraints` in solvers that only implement `copy_to_and_optimize!` since it's not even possible. ECOS now only stores the results so there is no way to implement it. These solvers will be used with a CachingOptimizer most of the time so there is no need to implement model-reading features.

Reply:
Then we don't really have a minimal API that every solver needs to implement. It's really "it depends."

Reply:
If we want to give up on defining a required API, then we at least should document that we intentionally use duck typing, and provide a list of recommended minimal attributes to support and a guide to the effect of, if you want your solver to support X then you need to implement Y.

Reply:
I think the main use cases are:
1) Be useable from JuMP
2) Be useable from JuMP in direct mode
3) Being able to copy from it
4) Being able to add things incrementally

and 2) = 1) + 4) (+ 3) if you want printing, etc... to work)